clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```
@@ -66,17 +69,102 @@
     private static final int NO_INFLIGHT_REQUEST_CORRELATION_ID = -1;
+    private static final int NO_LAST_ACKED_SEQUENCE_NUMBER = -1;
 
     private final Logger log;
     private final String transactionalId;
     private final int transactionTimeoutMs;
 
-    // The base sequence of the next batch bound for a given partition.
-    private final Map<TopicPartition, Integer> nextSequence;
+    private static class TopicPartitionBookkeeper {
 
-    // The sequence of the last record of the last ack'd batch from the given partition. When there are no
-    // in flight requests for a partition, the lastAckedSequence(topicPartition) == nextSequence(topicPartition) - 1.
-    private final Map<TopicPartition, Integer> lastAckedSequence;
+        private final Map<TopicPartition, TopicPartitionEntry> topicPartitionBookkeeping = new HashMap<>();
+
+        boolean hasPendingOffsetCommits() {
+            return topicPartitionBookkeeping.values().stream().anyMatch(tpe -> tpe.pendingTxnOffsetCommit != null);
+        }
+
+        void clearPendingOffsetCommits() {
+            topicPartitionBookkeeping.values().forEach(tpe -> tpe.pendingTxnOffsetCommit = null);
+        }
+
+        void clearPendingOffsetCommit(TopicPartition topic) {
+            TopicPartitionEntry tpe = topicPartitionBookkeeping.get(topic);
+            if (tpe != null) {
+                tpe.pendingTxnOffsetCommit = null;
+            }
+        }
+
+        Map<TopicPartition, CommittedOffset> pendingOffsetCommits() {
+            return topicPartitionBookkeeping.entrySet()
+                    .stream()
+                    .filter(e -> e.getValue() != null && e.getValue().pendingTxnOffsetCommit != null)
+                    .collect(Collectors.toMap(Map.Entry::getKey, t -> t.getValue().pendingTxnOffsetCommit));
+        }
+
+        public TopicPartitionEntry get(TopicPartition topic) {
+            TopicPartitionEntry ent = topicPartitionBookkeeping.get(topic);
+            if (ent == null) {
```

One thing I was considering is whether we should be explicit about when we expect these entries to be inserted. We could raise an exception from `get` here in cases where we expect the state to exist. For example, for the transactional producer, we have `maybeAddPartitionToTransaction` where we could explicitly insert. For idempotent producer, probably the first time we try to access the sequence for a partition or something. Anyway, just a thought, but it might make the code a little easier to reason about.

---

Basically I think I had that hasSequenceNumber stuff to do this but maybe it would be better if we raise an exception and add an insert method to assert the expected behavior as you suggested. I'll modify the code.

---

Did these changes and I think it has changed for the better.